appshell/appshell_extensions_gtk.cpp

```diff
@@ -448,18 +479,49 @@ int ReadFile(ExtensionString filename, ExtensionString encoding, std::string& co
             error = ERR_CANT_READ;
         }
     } else {
-        if (has_utf16_32_BOM(file_get_contents, len)) {
+        if (has_utf_32_BOM(file_get_contents, len)) {
             error = ERR_UNSUPPORTED_ENCODING;
         } else  if (has_utf8_BOM(file_get_contents, len)) {
             contents.assign(file_get_contents + utf8_BOM_Len, len);        
-        } else if (!g_locale_to_utf8(file_get_contents, -1, NULL, NULL, &gerror)) {
-            error = ERR_UNSUPPORTED_ENCODING;
+        } else if (!g_locale_to_utf8(file_get_contents, -1, NULL, NULL, &gerror) || encoding != "UTF-8") {
+            contents.assign(file_get_contents, len);
+            std::string detectedCharSet;
+            try {
+                if (encoding == "UTF-8") {
+                    GetCharsetMatch(contents.c_str(), contents.size(), detectedCharSet);
+                }
+                else {
+                    detectedCharSet = encoding;
+                }
+                std::transform(detectedCharSet.begin(), detectedCharSet.end(), detectedCharSet.begin(), ::toupper);
```

Comment:
This snippet could be made common across platforms. It will avoid repetition and is a good candidate for refactoring.